compendium/DeclarativeServices/src/manager/BindingPolicyDynamicGreedy.cpp

```diff
@@ -67,11 +68,23 @@ namespace cppmicroservices
                     {
                         notifications.emplace_back(mgr.metadata_.name, RefEvent::REBIND, reference);
                     }
-                    else
-                    { // there are bound refs, determine whether to rebind
+                    // there are bound refs, determine whether to rebind based on multiplicity of reference
+                    else if (mgr.IsUnary()) {
+                        //rebind new and unbind older reference if maxCardinality is unary
                         svcRefToUnBind = *(boundRefsHandle->begin());
                         needRebind = svcRefToUnBind < reference;
                     }
+                    else if (mgr.IsMultiple()) {
+                        //bind to new reference if maxCardinality is multiple
+                        if (boundRefsHandle->size() < mgr.metadata_.maxCardinality) {
+                            //number of bound references are within maxCardinality limit
+                            notifications.emplace_back(mgr.metadata_.name, RefEvent::REBIND, reference);
+                        }
+                        else {
+                            //Maximum limit of multiple references has reached, no new references will be bound
+                            Log("Number of multiple references has reached its maximum limit. New reference(s) will not be bound.", SeverityLevel::LOG_WARNING);
```

Comment:
What if we've reached maxCardinality but the new service is a better match than one that is already bound?

Reply:
My understanding was that in case of multiple cardinality we bind to any new reference irrespective of it being "better" or not, i.e. we are not comparing the rankings here. So in that case this logic would not bind to the new reference if the limit is reached.

Reply:
yes, that's my understanding, I interpret this table from the [OSGi spec](https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.component.html#service.component-reference.policy) to mean that for dynamic greedy, any new service is bound:

<html>
<body>
<!--StartFragment--><p class="title" style="box-sizing: border-box; font-weight: bold; font-style: italic; color: rgb(51, 51, 51); font-family: Ubuntu, Arial, Helvetica, sans-serif; font-size: 16px; font-variant-ligatures: normal; font-variant-caps: normal; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;"><strong style="box-sizing: border-box;">Table <span xmlns="" class="number" style="box-sizing: border-box;">112</span>.1 Action taken for policy-option when a new or higher ranking service becomes available</strong></p><div class="table-contents" style="box-sizing: border-box; overflow-x: auto; color: rgb(51, 51, 51); font-family: Ubuntu, Arial, Helvetica, sans-serif; font-size: 16px; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; background-color: rgb(255, 255, 255); text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">

Cardinality | static reluctant | static greedy | dynamic reluctant | dynamic greedy
-- | -- | -- | -- | --
0..1 | Ignore | Reactivate to bind the better target service. | If no service is bound, bind to new target service. Otherwise, ignore new target service. | If no service is bound, bind to better target service. Otherwise, unbind the bound service and bind the better target service.
1..1 | Ignore | Reactivate to bind the better target service. | Ignore | Unbind the bound service, then bind the new service.
0..n | Ignore | Reactivate | Bind new target service | Bind new target service
1..n | Ignore | Reactivate | Bind new target service | Bind new target service

</div><!--EndFragment-->
</body>
</html>